src/Interpreters/SessionLog.cpp

```diff
@@ -212,8 +212,8 @@ void SessionLog::addLoginSuccess(const UUID & auth_id, std::optional<String> ses
     DB::SessionLogElement log_entry(auth_id, SESSION_LOGIN_SUCCESS);
     log_entry.client_info = client_info;
 
+    if (const auto user = login_context.getUser())
```

Comment:
Why can it be nullptr? Seems like we cannot have a session without user.

Reply:
+1, if login sucessful, we should expect the user exists.

Reply:
But this somehow happened in the test.

`Context::getUser()` is equivalent to `context.getAccess()->getUser()`:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Context.cpp#L750

In `Session::makeQueryContextImpl` we check that `user` is non-null before invoking `SessionLog::addLoginSuccess`: https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Session.cpp#L484

Yet somehow it became null in `SessionLog::addLoginSuccess` in the [test](https://github.com/ClickHouse/ClickHouse/issues/35714).

The context we are working on is local, and can't be shared between multiple threads, but `ContextAccess` is shared between contexts (e.g. original session context). It's `ContextAccess::user` and it can be changed and set to null in a change-handling callback:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Access/ContextAccess.cpp#L163

Reply:
Can it be a race condition with `DROP USER`, for example? Need to check server logs. In this case on successful login we should save shared pointer to `user` somewhere to be able to write meaningful row into session log. Imagine that you are checking login attempts in session log on your server and see some successful logins without information about user... I suppose that it will really scare you, because it looks like someone has hacked your server.

Reply:
I didn't find the corresponding `DROP USER` query in server logs, so it's still not clear what happened.